### PR TITLE
video_core: Generate mipmap texture by drawing

### PR DESCRIPTION
**Pull request by vonchenplus**
It seems that some games use drawing to generate mipmap textures, and the emulator does not currently support this behaviour very well. 

When this behaviour is triggered, the viewport transform is disabled, so we can set the current viewport to the drawing area and rescale the vertex position.

image
To
image

Fixed：
Icons in RetroArch
image
image

DQB sees similar behavior, but needs more fixes.

**Warning: This will broken the existing shader cache.**

### CODE REVIEW

File: src/shader_recompiler/ir_opt/position_pass.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+namespace Shader::Optimization {
+
+namespace {
+struct PositionInst {
+    IR::Inst* inst;
+    IR::Block* block;
+    IR::Attribute attr;
+};
+using PositionInstVector = boost::container::small_vector<PositionInst, 24>;
```
Comment (lioncash) [inline src/shader_recompiler/ir_opt/position_pass.cpp:17]: the small vector header should be included in this file and not be relying on indirect inclusions

File: src/video_core/renderer_opengl/gl_rasterizer.cpp [hunk 2]
```diff
@@ -622,6 +622,16 @@ void RasterizerOpenGL::SyncViewport() {
             }
             flags[Dirty::Viewport0 + index] = false;
 
+            if (!regs.viewport_transform_enabled) {
+                GLfloat x = static_cast<float>(regs.render_area.x);
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_rasterizer.cpp:626]:
```suggestion
                const auto x = static_cast<GLfloat>(regs.render_area.x);
```

Ditto for the ones below

File: src/shader_recompiler/ir_opt/position_pass.cpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+} // namespace
```
Comment (Morph1984) [inline src/shader_recompiler/ir_opt/position_pass.cpp:20]: // Anonymous namespace

File: src/shader_recompiler/ir_opt/position_pass.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+void PositionPass(Environment& env, IR::Program& program) {
+    if (env.ShaderStage() != Stage::VertexB || env.ReadViewportTransformState())
+        return;
```
Comment (Morph1984) [inline src/shader_recompiler/ir_opt/position_pass.cpp:24]:
```cpp
if (...) {
    return
}
```

File: src/shader_recompiler/ir_opt/position_pass.cpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                1,
+                ir.FPFma(value, ir.FPMul(ir.FPRecip(ir.RenderAreaHeight()), scale), negative_one));
+            break;
+        }
+        default:
+            break;
+        }
+    }
+}
+} // namespace Shader::Optimization
```
Comment (Morph1984) [inline src/shader_recompiler/ir_opt/position_pass.cpp:76]: add newline after this line

File: src/video_core/renderer_vulkan/vk_graphics_pipeline.cpp [hunk 6]
```diff
@@ -483,11 +494,17 @@ void GraphicsPipeline::ConfigureDraw(const RescalingPushConstant& rescaling) {
...
             cmdbuf.PushConstants(*pipeline_layout, VK_SHADER_STAGE_ALL_GRAPHICS,
                                  RESCALING_LAYOUT_DOWN_FACTOR_OFFSET, sizeof(scale_down_factor),
                                  &scale_down_factor);
         }
+        if (uses_render_area) {
+            cmdbuf.PushConstants(*pipeline_layout, VK_SHADER_STAGE_ALL_GRAPHICS,
+                                 RENDERAREA_LAYOUT_OFFSET, sizeof(render_area_data),
+                                 &render_area_data);
+        }
+
```
Comment (Morph1984) [inline src/video_core/renderer_vulkan/vk_graphics_pipeline.cpp:507]: remove this newline (inconsistent with above and below)

File: src/shader_recompiler/CMakeLists.txt [hunk 7]
```diff
@@ -228,6 +228,7 @@ add_library(shader_recompiler STATIC
     ir_opt/ssa_rewrite_pass.cpp
     ir_opt/texture_pass.cpp
     ir_opt/verification_pass.cpp
+    ir_opt/position_pass.cpp
```
Comment (Morph1984) [inline src/shader_recompiler/CMakeLists.txt:231]: move this between passes.h and rescaling_pass.cpp

File: src/shader_recompiler/backend/spirv/spirv_emit_context.cpp [hunk 8]
```diff
@@ -962,6 +963,7 @@ void EmitContext::DefineRescalingInputPushConstant() {
                        static_cast<u32>(offsetof(RescalingLayout, down_factor)));
         MemberName(push_constant_struct, rescaling_downfactor_member_index, "down_factor");
     }
+
```
Comment (Morph1984) [inline src/shader_recompiler/backend/spirv/spirv_emit_context.cpp:966]: remove newline (inconsistent with above)

File: src/shader_recompiler/backend/spirv/spirv_emit_context.h [hunk 9]
```diff
@@ -243,6 +243,9 @@ class EmitContext final : public Sirit::Module {
     u32 texture_rescaling_index{};
     u32 image_rescaling_index{};
 
+    Id render_area_push_constant;
+    u32 render_are_member_index;
```
Comment (Morph1984) [inline src/shader_recompiler/backend/spirv/spirv_emit_context.h:247]:
```suggestion
    Id render_area_push_constant{};
    u32 render_area_member_index{};
```

**Comment by Timmy5942** [comment]
Rebase?

**Review by bunnei — APPROVED** [review]

**Pull request merged**